[BUG] Can not install local package in a workspace. Local package name and version affects resolution.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

Create a folder and run:
```bash
npm init -y
npm init -w packages/a -y
npm init -w packages/b -y
npm init -w packages/c -y
npm init -w packages/z -y
npm i b@1.0.0 -w a
npm i c@1.0.0 -w a
npm i z@1.0.0 -w a
```

Every line completes without errors except last one. At the end you will get:
```
npm ERR! code ETARGET
npm ERR! notarget No matching version found for z@1.0.0.
npm ERR! notarget In most cases you or one of your dependencies are requesting
npm ERR! notarget a package version that doesn't exist.
```

Local packages `b` and `c` were successfully added to the local package `a`. Adding package `z` raised error.
#### It turns out that if there is no package with such name and version in the **nmp registry,** it will fail to install **local package**.


### Expected Behavior

I should be able to install any local packages inside my **workspace** regardless of remote package with the same name and version being or NOT being presented in the remote npm repository.   

### Steps To Reproduce

1. In an empty folder run:
```
npm init -y
npm init -w packages/a -y
npm init -w packages/b -y
npm init -w packages/c -y
npm init -w packages/z -y
npm i b@1.0.0 -w a
npm i c@1.0.0 -w a
npm i z@1.0.0 -w a
```

### Environment

- npm: 8.1.2, 8.1.4
- Node.js: v16.13.1, v17.0.1
- OS Name: macOS Monterey 
- System Model Name: Macbook Pro M1 
- npm config:
```ini
; node bin location = /Users/user/.nvm/versions/node/v17.0.1/bin/node
; cwd = /Users/user/Projects/delete_me/npm_workspaces_problem
; HOME = /Users/user
; Run `npm config ls -l` to show all defaults.
; "default" config from default values

_auth = (protected) 
access = null 
all = false 
allow-same-version = false 
also = null 
audit = true 
audit-level = null 
auth-type = "legacy" 
before = null 
bin-links = true 
browser = null 
ca = null 
cache = "/Users/user/.npm" 
cache-max = null 
cache-min = 0 
cafile = null 
call = "" 
cert = null 
ci-name = null 
cidr = null 
color = true 
commit-hooks = true 
depth = null 
description = true 
dev = false 
diff = [] 
diff-dst-prefix = "b/" 
diff-ignore-all-space = false 
diff-name-only = false 
diff-no-prefix = false 
diff-src-prefix = "a/" 
diff-text = false 
diff-unified = 3 
dry-run = false 
editor = "vi" 
engine-strict = false 
fetch-retries = 2 
fetch-retry-factor = 10 
fetch-retry-maxtimeout = 60000 
fetch-retry-mintimeout = 10000 
fetch-timeout = 300000 
force = false 
foreground-scripts = false 
format-package-lock = true 
fund = true 
git = "git" 
git-tag-version = true 
global = false 
global-style = false 
globalconfig = "/Users/user/.nvm/versions/node/v17.0.1/etc/npmrc" 
heading = "npm" 
https-proxy = null 
if-present = false 
ignore-scripts = false 
include = [] 
include-staged = false 
include-workspace-root = false 
init-author-email = "" 
init-author-name = "" 
init-author-url = "" 
init-license = "ISC" 
init-module = "/Users/user/.npm-init.js" 
init-version = "1.0.0" 
init.author.email = "" 
init.author.name = "" 
init.author.url = "" 
init.license = "ISC" 
init.module = "/Users/user/.npm-init.js" 
init.version = "1.0.0" 
json = false 
key = null 
legacy-bundling = false 
legacy-peer-deps = false 
link = false 
local-address = null 
location = "user" 
lockfile-version = null 
loglevel = "notice" 
logs-max = 10 
; long = false ; overridden by cli
maxsockets = 15 
message = "%s" 
metrics-registry = "https://registry.npmjs.org/" 
node-options = null 
node-version = "v17.0.1" 
noproxy = [""] 
npm-version = "8.1.4" 
offline = false 
omit = [] 
only = null 
optional = null 
otp = null 
pack-destination = "." 
package = [] 
package-lock = true 
package-lock-only = false 
parseable = false 
prefer-offline = false 
prefer-online = false 
prefix = "/Users/user/.nvm/versions/node/v17.0.1" 
preid = "" 
production = null 
progress = true 
proxy = null 
read-only = false 
rebuild-bundle = true 
registry = "https://registry.npmjs.org/" 
save = true 
save-bundle = false 
save-dev = false 
save-exact = false 
save-optional = false 
save-peer = false 
save-prefix = "^" 
save-prod = false 
scope = "" 
script-shell = null 
searchexclude = "" 
searchlimit = 20 
searchopts = "" 
searchstaleness = 900 
shell = "/bin/zsh" 
shrinkwrap = true 
sign-git-commit = false 
sign-git-tag = false 
sso-poll-frequency = 500 
sso-type = "oauth" 
strict-peer-deps = false 
strict-ssl = true 
tag = "latest" 
tag-version-prefix = "v" 
timing = false 
tmp = "/var/folders/gb/40nmdn5d45b_98f3p6ppwwmr0000gn/T" 
umask = 0 
unicode = true 
update-notifier = true 
usage = false 
user-agent = "npm/{npm-version} node/{node-version} {platform} {arch} workspaces/{workspaces} {ci}" 
userconfig = **"/Users/user/.npmrc"** 
version = false 
versions = false 
viewer = "man" 
which = null 
workspace = [] 
workspaces = null 
yes = null 

; "cli" config from command line options

long = true

```

--- a/Makefile
+++ b/Makefile
@@ -85,7 +85,7 @@ test-all: deps
 	node bin/npm-cli.js run test-all
 
 ls-ok:
-	node bin/npm-cli.js ls --production >/dev/null
+	node bin/npm-cli.js ls --omit=dev >/dev/null
 
 gitclean:
 	git clean -fd
@@ -97,7 +97,7 @@ link: uninstall
 	node bin/npm-cli.js link -f --ignore-scripts
 
 prune: deps
-	node bin/npm-cli.js prune --production --no-save --no-audit --no-fund
+	node bin/npm-cli.js prune --omit=dev --no-save --no-audit --no-fund
 	node scripts/git-dirty.js
 
 publish: gitclean ls-ok link test-all docs prune

--- a/lib/commands/ls.js
+++ b/lib/commands/ls.js
@@ -52,16 +52,12 @@ class LS extends ArboristWorkspaceCmd {
     const all = this.npm.config.get('all')
     const color = this.npm.color
     const depth = this.npm.config.get('depth')
-    const dev = this.npm.config.get('dev')
-    const development = this.npm.config.get('development')
     const global = this.npm.config.get('global')
     const json = this.npm.config.get('json')
     const link = this.npm.config.get('link')
     const long = this.npm.config.get('long')
-    const only = this.npm.config.get('only')
+    const omit = this.npm.flatOptions.omit
     const parseable = this.npm.config.get('parseable')
-    const prod = this.npm.config.get('prod')
-    const production = this.npm.config.get('production')
     const unicode = this.npm.config.get('unicode')
     const packageLockOnly = this.npm.config.get('package-lock-only')
     const workspacesEnabled = this.npm.flatOptions.workspacesEnabled
@@ -138,15 +134,10 @@ class LS extends ArboristWorkspaceCmd {
           ? []
           : [...(node.target).edgesOut.values()]
             .filter(filterBySelectedWorkspaces)
-            .filter(filterByEdgesTypes({
-              currentDepth,
-              dev,
-              development,
+            .filter(currentDepth === 0 ? filterByEdgesTypes({
               link,
-              prod,
-              production,
-              only,
-            }))
+              omit,
+            }) : () => true)
             .map(mapEdgesToNodes({ seenPaths }))
             .concat(appendExtraneousChildren({ node, seenPaths }))
             .sort(sortAlphabetically)
@@ -399,27 +390,13 @@ const getJsonOutputItem = (node, { global, long }) => {
   return augmentItemWithIncludeMetadata(node, item)
 }
 
-const filterByEdgesTypes = ({
-  currentDepth,
-  dev,
-  development,
-  link,
-  prod,
-  production,
-  only,
-}) => {
-  // filter deps by type, allows for: `npm ls --dev`, `npm ls --prod`,
-  // `npm ls --link`, `npm ls --only=dev`, etc
-  const filterDev = currentDepth === 0 &&
-    (dev || development || /^dev(elopment)?$/.test(only))
-  const filterProd = currentDepth === 0 &&
-    (prod || production || /^prod(uction)?$/.test(only))
-  const filterLink = currentDepth === 0 && link
-
-  return (edge) =>
-    (filterDev ? edge.dev : true) &&
-    (filterProd ? (!edge.dev && !edge.peer && !edge.peerOptional) : true) &&
-    (filterLink ? (edge.to && edge.to.isLink) : true)
+const filterByEdgesTypes = ({ link, omit = [] }) => (edge) => {
+  for (const omitType of omit) {
+    if (edge[omitType]) {
+      return false
+    }
+  }
+  return link ? edge.to && edge.to.isLink : true
 }
 
 const appendExtraneousChildren = ({ node, seenPaths }) =>

--- a/lib/commands/outdated.js
+++ b/lib/commands/outdated.js
@@ -208,7 +208,7 @@ class Outdated extends ArboristWorkspaceCmd {
       : edge.dev ? 'devDependencies'
       : 'dependencies'
 
-    for (const omitType of this.npm.config.get('omit')) {
+    for (const omitType of this.npm.flatOptions.omit) {
       if (node[omitType]) {
         return
       }

--- a/test/lib/commands/ls.js
+++ b/test/lib/commands/ls.js
@@ -99,14 +99,12 @@ const LS = t.mock('../../../lib/commands/ls.js', {
 const config = {
   all: true,
   color: false,
-  dev: false,
   depth: Infinity,
   global: false,
   json: false,
   link: false,
-  only: null,
+  omit: [],
   parseable: false,
-  production: false,
   'package-lock-only': false,
 }
 const flatOptions = {
@@ -456,7 +454,7 @@ t.test('ls', t => {
   })
 
   t.test('--dev', async t => {
-    config.dev = true
+    flatOptions.omit = ['peer', 'prod', 'optional']
     npm.prefix = t.testdir({
       'package.json': JSON.stringify({
         name: 'test-npm-ls',
@@ -479,34 +477,7 @@ t.test('ls', t => {
     })
     await ls.exec([])
     t.matchSnapshot(redactCwd(result), 'should output tree containing dev deps')
-    config.dev = false
-  })
-
-  t.test('--only=development', async t => {
-    config.only = 'development'
-    npm.prefix = t.testdir({
-      'package.json': JSON.stringify({
-        name: 'test-npm-ls',
-        version: '1.0.0',
-        dependencies: {
-          'prod-dep': '^1.0.0',
-          chai: '^1.0.0',
-        },
-        devDependencies: {
-          'dev-dep': '^1.0.0',
-        },
-        optionalDependencies: {
-          'optional-dep': '^1.0.0',
-        },
-        peerDependencies: {
-          'peer-dep': '^1.0.0',
-        },
-      }),
-      ...diffDepTypesNmFixture,
-    })
-    await ls.exec([])
-    t.matchSnapshot(redactCwd(result), 'should output tree containing only development deps')
-    config.only = null
+    flatOptions.omit = []
   })
 
   t.test('--link', async t => {
@@ -581,7 +552,7 @@ t.test('ls', t => {
   })
 
   t.test('--production', async t => {
-    config.production = true
+    flatOptions.omit = ['dev', 'peer']
     npm.prefix = t.testdir({
       'package.json': JSON.stringify({
         name: 'test-npm-ls',
@@ -604,34 +575,7 @@ t.test('ls', t => {
     })
     await ls.exec([])
     t.matchSnapshot(redactCwd(result), 'should output tree containing production deps')
-    config.production = false
-  })
-
-  t.test('--only=prod', async t => {
-    config.only = 'prod'
-    npm.prefix = t.testdir({
-      'package.json': JSON.stringify({
-        name: 'test-npm-ls',
-        version: '1.0.0',
-        dependencies: {
-          'prod-dep': '^1.0.0',
-          chai: '^1.0.0',
-        },
-        devDependencies: {
-          'dev-dep': '^1.0.0',
-        },
-        optionalDependencies: {
-          'optional-dep': '^1.0.0',
-        },
-        peerDependencies: {
-          'peer-dep': '^1.0.0',
-        },
-      }),
-      ...diffDepTypesNmFixture,
-    })
-    await ls.exec([])
-    t.matchSnapshot(redactCwd(result), 'should output tree containing only prod deps')
-    config.only = null
+    flatOptions.omit = []
   })
 
   t.test('--long', async t => {
@@ -1484,12 +1428,12 @@ t.test('ls', t => {
     t.matchSnapshot(redactCwd(result), 'should list --all workspaces properly')
 
     // --production
-    config.production = true
+    flatOptions.omit = ['dev', 'peer', 'optional']
     await ls.exec([])
 
     t.matchSnapshot(redactCwd(result), 'should list only prod deps of workspaces')
 
-    config.production = false
+    flatOptions.omit = []
 
     // filter out a single workspace using args
     await ls.exec(['d'])
@@ -1811,7 +1755,7 @@ t.test('ls --parseable', t => {
   })
 
   t.test('--dev', async t => {
-    config.dev = true
+    flatOptions.omit = ['peer', 'prod', 'optional']
     npm.prefix = t.testdir({
       'package.json': JSON.stringify({
         name: 'test-npm-ls',
@@ -1834,34 +1778,7 @@ t.test('ls --parseable', t => {
     })
     await ls.exec([])
     t.matchSnapshot(redactCwd(result), 'should output tree containing dev deps')
-    config.dev = false
-  })
-
-  t.test('--only=development', async t => {
-    config.only = 'development'
-    npm.prefix = t.testdir({
-      'package.json': JSON.stringify({
-        name: 'test-npm-ls',
-        version: '1.0.0',
-        dependencies: {
-          'prod-dep': '^1.0.0',
-          chai: '^1.0.0',
-        },
-        devDependencies: {
-          'dev-dep': '^1.0.0',
-        },
-        optionalDependencies: {
-          'optional-dep': '^1.0.0',
-        },
-        peerDependencies: {
-          'peer-dep': '^1.0.0',
-        },
-      }),
-      ...diffDepTypesNmFixture,
-    })
-    await ls.exec([])
-    t.matchSnapshot(redactCwd(result), 'should output tree containing only development deps')
-    config.only = null
+    flatOptions.omit = []
   })
 
   t.test('--link', async t => {
@@ -1902,7 +1819,7 @@ t.test('ls --parseable', t => {
   })
 
   t.test('--production', async t => {
-    config.production = true
+    flatOptions.omit = ['dev', 'peer']
     npm.prefix = t.testdir({
       'package.json': JSON.stringify({
         name: 'test-npm-ls',
@@ -1925,34 +1842,7 @@ t.test('ls --parseable', t => {
     })
     await ls.exec([])
     t.matchSnapshot(redactCwd(result), 'should output tree containing production deps')
-    config.production = false
-  })
-
-  t.test('--only=prod', async t => {
-    config.only = 'prod'
-    npm.prefix = t.testdir({
-      'package.json': JSON.stringify({
-        name: 'test-npm-ls',
-        version: '1.0.0',
-        dependencies: {
-          'prod-dep': '^1.0.0',
-          chai: '^1.0.0',
-        },
-        devDependencies: {
-          'dev-dep': '^1.0.0',
-        },
-        optionalDependencies: {
-          'optional-dep': '^1.0.0',
-        },
-        peerDependencies: {
-          'peer-dep': '^1.0.0',
-        },
-      }),
-      ...diffDepTypesNmFixture,
-    })
-    await ls.exec([])
-    t.matchSnapshot(redactCwd(result), 'should output tree containing only prod deps')
-    config.only = null
+    flatOptions.omit = []
   })
 
   t.test('--long', async t => {
@@ -2935,7 +2825,7 @@ t.test('ls --json', t => {
   })
 
   t.test('--dev', async t => {
-    config.dev = true
+    flatOptions.omit = ['prod', 'optional', 'peer']
     npm.prefix = t.testdir({
       'package.json': JSON.stringify({
         name: 'test-npm-ls',
@@ -2976,52 +2866,7 @@ t.test('ls --json', t => {
       },
       'should output json containing dev deps'
     )
-    config.dev = false
-  })
-
-  t.test('--only=development', async t => {
-    config.only = 'development'
-    npm.prefix = t.testdir({
-      'package.json': JSON.stringify({
-        name: 'test-npm-ls',
-        version: '1.0.0',
-        dependencies: {
-          'prod-dep': '^1.0.0',
-          chai: '^1.0.0',
-        },
-        devDependencies: {
-          'dev-dep': '^1.0.0',
-        },
-        optionalDependencies: {
-          'optional-dep': '^1.0.0',
-        },
-        peerDependencies: {
-          'peer-dep': '^1.0.0',
-        },
-      }),
-      ...diffDepTypesNmFixture,
-    })
-    await ls.exec([])
-    t.same(
-      jsonParse(result),
-      {
-        name: 'test-npm-ls',
-        version: '1.0.0',
-        dependencies: {
-          'dev-dep': {
-            version: '1.0.0',
-            dependencies: {
-              foo: {
-                version: '1.0.0',
-                dependencies: { dog: { version: '1.0.0' } },
-              },
-            },
-          },
-        },
-      },
-      'should output json containing only development deps'
-    )
-    config.only = null
+    flatOptions.omit = []
   })
 
   t.test('--link', async t => {
@@ -3075,7 +2920,7 @@ t.test('ls --json', t => {
   })
 
   t.test('--production', async t => {
-    config.production = true
+    flatOptions.omit = ['dev', 'peer']
     npm.prefix = t.testdir({
       'package.json': JSON.stringify({
         name: 'test-npm-ls',
@@ -3110,46 +2955,7 @@ t.test('ls --json', t => {
       },
       'should output json containing production deps'
     )
-    config.production = false
-  })
-
-  t.test('--only=prod', async t => {
-    config.only = 'prod'
-    npm.prefix = t.testdir({
-      'package.json': JSON.stringify({
-        name: 'test-npm-ls',
-        version: '1.0.0',
-        dependencies: {
-          'prod-dep': '^1.0.0',
-          chai: '^1.0.0',
-        },
-        devDependencies: {
-          'dev-dep': '^1.0.0',
-        },
-        optionalDependencies: {
-          'optional-dep': '^1.0.0',
-        },
-        peerDependencies: {
-          'peer-dep': '^1.0.0',
-        },
-      }),
-      ...diffDepTypesNmFixture,
-    })
-    await ls.exec([])
-    t.same(
-      jsonParse(result),
-      {
-        name: 'test-npm-ls',
-        version: '1.0.0',
-        dependencies: {
-          chai: { version: '1.0.0' },
-          'optional-dep': { version: '1.0.0' },
-          'prod-dep': { version: '1.0.0', dependencies: { dog: { version: '2.0.0' } } },
-        },
-      },
-      'should output json containing only prod deps'
-    )
-    config.only = null
+    flatOptions.omit = []
   })
 
   t.test('from lockfile', async t => {

--- a/test/lib/commands/outdated.js
+++ b/test/lib/commands/outdated.js
@@ -84,10 +84,6 @@ const globalDir = t.testdir({
   },
 })
 
-const flatOptions = {
-  workspacesEnabled: true,
-}
-
 const outdated = (dir, opts) => {
   logs = ''
   const Outdated = t.mock('../../../lib/commands/outdated.js', {
@@ -95,11 +91,22 @@ const outdated = (dir, opts) => {
       packument,
     },
   })
+  if (opts.config && opts.config.omit) {
+    opts.flatOptions = {
+      omit: opts.config.omit,
+      ...opts.flatOptions,
+    }
+    delete opts.config.omit
+  }
   const npm = mockNpm({
     ...opts,
     localPrefix: dir,
     prefix: dir,
-    flatOptions,
+    flatOptions: {
+      workspacesEnabled: true,
+      omit: [],
+      ...opts.flatOptions,
+    },
     globalDir: `${globalDir}/node_modules`,
     output,
   })
@@ -529,14 +536,16 @@ t.test('workspaces', async t => {
   t.matchSnapshot(logs, 'should display ws outdated deps human output')
   t.equal(process.exitCode, 1)
 
-  flatOptions.workspacesEnabled = false
-  await outdated(testDir, {}).exec([])
+  await outdated(testDir, {
+    flatOptions: {
+      workspacesEnabled: false,
+    },
+  }).exec([])
 
   // TODO: This should display dog, but doesn't because arborist filters
   // workspace deps even if they're also root deps
   // This will be fixed in a future arborist version
   t.matchSnapshot(logs, 'should display only root outdated when ws disabled')
-  flatOptions.workspacesEnabled = true
 
   await outdated(testDir, {
     config: {

